vdpa/virtio: Add support for virtio-blk for vdpa driver

kailiangz1 opened the pull request:
Enable config_get/set feature for virtio-blk
Add config_get/set vhost message handle in vhost framework

Signed-off-by: Kailiang Zhou <kailiangz@nvidia.com>

paravmellanox commented on lib/vhost/vhost_user.h:
```diff
@@ -125,6 +128,18 @@ typedef struct VhostUserInflight {
+#define VHOST_USER_MAX_CONFIG_SIZE		256
+#endif
+
+/** Get/set config msg payload */
```
remove one extra "*".

paravmellanox commented on lib/vhost/vhost_user.c:
```diff
@@ -3311,7 +3355,7 @@ int rte_vhost_host_notifier_ctrl(int vid, uint16_t qid, bool enable)
 
 	if (!(dev->features & (1ULL << VIRTIO_F_VERSION_1)) ||
-	    !(dev->features & (1ULL << VHOST_USER_F_PROTOCOL_FEATURES)) ||
+/*	    !(dev->features & (1ULL << VHOST_USER_F_PROTOCOL_FEATURES)) ||-----blk device will not negotiate protocal feature, but still can use notify*/
```
too long line, cut it out.

paravmellanox commented on drivers/common/virtio/virtio_blk.h:
```diff
@@ -0,0 +1,177 @@
+ * For legacy virtio, if VIRTIO_F_ANY_LAYOUT is not negotiated,
+ * this is the first element of the read scatter-gather list.
+ */
+struct virtio_blk_outhdr {
```
remove the code that you are not using here.

GavinLi-NV commented on drivers/common/virtio/virtio_blk.h:
```diff
@@ -0,0 +1,177 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright (c) 2022 NVIDIA Corporation & Affiliates
```
Please also include this file in virtio.h, just like virtio_net.h

paravmellanox commented on drivers/common/virtio/virtio_pci.c:
```diff
@@ -600,6 +617,10 @@ const struct virtio_dev_specific_ops virtio_net_dev_pci_modern_ops = {
 };
 
+const struct virtio_dev_specific_ops virtio_blk_dev_pci_modern_ops = {
+	.get_queue_num	= modern_blk_get_queue_num,
```
no need for extra white space after get_queue_num. remove.

paravmellanox commented on examples/vdpa/main.c:
```diff
@@ -179,6 +180,19 @@ start_vdpa(struct vdpa_port *vport)
 			socket_path);
 
+	/*vdpa device should get feature from device and not use builtin_net_driver*/
+	ret = rte_vdpa_get_features(vport->dev,&features);
```
white space after dev, is needed.

paravmellanox commented on examples/vdpa/main.c:
```diff
@@ -179,6 +180,19 @@ start_vdpa(struct vdpa_port *vport)
 			"register driver failed: %s\n",
 			socket_path);
 
+	/*vdpa device should get feature from device and not use builtin_net_driver*/
```
white space before and after /* and */ is needed.

you need to review those cosmetics once before posting here.

paravmellanox commented on lib/vhost/vhost_user.c:
```diff
@@ -2131,6 +2168,11 @@ vhost_user_get_vring_base(struct virtio_net **pdev,
 	/* We have to stop the queue (virtio) if it is running. */
 	vhost_destroy_device_notify(dev);
 
+	/* Device is not running after get vring base, we should unmask driver_ok.
```
all of these changes for get and set config etc, has nothing to do with blk.

split the patches to two.
one for blk actual work and one for this.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,69 @@
+};
+
+#define VIRTIO_VDPA_DRIVER_NAME vdpa_virtio
+#define VIRTIO_VDPA_NET_PROTOCOL_FEATURES \
```
this line doesnt belong to this generic virtio.h keep in the net file.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -138,7 +88,15 @@ virtio_vdpa_protocol_features_get(struct rte_vdpa_device *vdev,
+	if (priv->pdev->id.device_id == VIRTIO_PCI_MODERN_DEVICEID_NET)
+		*features = VIRTIO_VDPA_NET_PROTOCOL_FEATURES;
+	else if (priv->pdev->id.device_id == VIRTIO_PCI_MODERN_DEVICEID_BLK)
+		*features = VIRTIO_VDPA_BLK_PROTOCOL_FEATURES;
```
get them from the device specific callback.

kailiangz1 committed:
vhost/vdpa: Add config get/set msg handler

For virito-blk device, vhost must support config msg.
So, add support for vhost framework

Signed-off-by: Kailiang Zhou <kailiangz@nvidia.com>

kailiangz1 committed:
vdpa/virtio: Add support for virtio-blk for vdpa driver

Enable config_get/set feature for virtio-blk
Seprate vhost feature get for blk and net device

Signed-off-by: Kailiang Zhou <kailiangz@nvidia.com>

GavinLi-NV commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -935,6 +902,11 @@ virtio_vdpa_dev_probe(struct rte_pci_driver *pci_drv __rte_unused,
 		goto error;
```
Mem leak

GavinLi-NV commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -935,6 +902,11 @@ virtio_vdpa_dev_probe(struct rte_pci_driver *pci_drv __rte_unused,
+
 	priv->vfio_dev_fd = rte_intr_dev_fd_get(pci_dev->intr_handle);
 	if (priv->vfio_dev_fd < 0) {
 		DRV_LOG(ERR, "%s failed to get vfio dev fd", devname);
```
After virtio_pci_dev_alloc, rte_vdpa_register_device, virtio_pci_dev_interrupts_alloc, mem leak will happen

paravmellanox commented on drivers/common/virtio/virtio_blk.h:
```diff
@@ -0,0 +1,141 @@
+ */
+
+/* These two define direction. */
+#define VIRTIO_BLK_T_IN		0
```
fix extra tab before 0

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,56 @@
+extern struct virtio_vdpa_device_callback virtio_vdpa_blk_callback;
+extern struct virtio_vdpa_device_callback virtio_vdpa_net_callback;
+
+#define VIRTIO_VDPA_INTR_RETRIES_USEC 1000
```
I think these two variables can stay back in the .c file.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,56 @@
+	bool configured;
+};
+
+extern struct virtio_vdpa_device_callback virtio_vdpa_blk_callback;
```
externs should be in .c file and not .h.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,56 @@
+	struct virtio_vdpa_priv *priv;
+};
+
+#define VIRTIO_VDPA_DRIVER_NAME vdpa_virtio
```
string should be in "vdpa_virtio".

kailiangz1 replied:
current in DPDK, all driver don't use "" in RTE_PMD_REGISTER_PCI

paravmellanox commented on drivers/common/virtio/virtio_pci.c:
```diff
@@ -468,6 +469,22 @@ modern_net_get_queue_num(struct virtio_hw *hw)
 }
 
+static uint16_t
+modern_blk_get_queue_num(struct virtio_hw *hw)
```
can this function be shifted to virtio_vdpa_blk.c similar for net.c

paravmellanox commented on drivers/common/virtio/virtio_pci_net.c:
```diff
@@ -0,0 +1,44 @@
+#include "virtqueue.h"
+
+static uint16_t
+modern_net_get_queue_num(struct virtio_hw *hw)
```
cut down this routine. it was overkill from the beginning.

struct virtio_pci_common_cfg {} structure has num_queues field for all the device types including net and blk.
This needs to reflect the maximum queue value regardless of device type.

Use it uniformly for net and blk.
and put in common code.

kailiangz1 replied:
num_queue include control queue and admin queue.
VDPA　driver use this to get how many IO queue device can support.
so, shoule only return number of IO queue.
So, blk and net is different also.
we should keep this code , but only remove admin queue and controler queue logic.

paravmellanox replied:
@kailiangz1 the logic should be in get_vqs_num() callbkc to be,

get_hw_num_vqs() -1 for net cvq.

kailiangz1 replied:
It should not like this, get_hw_num_vqs to get from common config's num_queues is hard to judge
whether the number include cvq or admin vq.
Directly get from virtio_net_config obey the spec and do need to calculate.

paravmellanox replied:
@kailiangz1 num_vqs is generic for all the protocols regardless of device type. This is the virtio pci layer provides.
vdpa net and block layer can derive its own value from low layer.

This is simple math. no need to judge/guess anything.

kailiangz1 replied:
we need to judge whether cvq is enabled and whether admin vq is enabled.

paravmellanox commented on examples/vdpa/main.c:
```diff
@@ -179,6 +180,19 @@ start_vdpa(struct vdpa_port *vport)
 
+	/* vdpa device should get feature from device and not use builtin_net_driver */
+	ret = rte_vdpa_get_features(vport->dev, &features);
+	if (ret != 0)
```
just do

if (ret)

below also

kailiangz1 replied:
all the other code also use this format, if change this,
it will not consistent with other code.

paravmellanox commented on examples/vdpa/main.c:
```diff
@@ -179,6 +180,19 @@ start_vdpa(struct vdpa_port *vport)
+			socket_path);
+
+	ret = rte_vhost_driver_set_features(socket_path, features);
+	if (ret != 0)
```
if (ret)

paravmellanox commented on examples/vdpa/main.c:
```diff
@@ -179,6 +180,19 @@ start_vdpa(struct vdpa_port *vport)
+
 	ret = rte_vhost_driver_callback_register(socket_path,
 			&vdpa_sample_devops);
 	if (ret != 0)
```
if (ret)

paravmellanox commented on lib/vhost/vhost_user.c:
```diff
@@ -342,6 +344,35 @@ vhost_user_get_features(struct virtio_net **pdev,
+	vdpa_dev = dev->vdpa_dev;
+
+	if (vdpa_dev)
+		vdpa_dev->ops->get_dev_config(dev->vid,ctx->msg.payload.cfg.region,ctx->msg.payload.cfg.size);
```
white space after, ","

split to two lines.

Please run kernel checkpatch.pl before pushing changes to catch such silly cosmetic changes.

paravmellanox commented on lib/vhost/vhost_user.c:
```diff
@@ -342,6 +344,35 @@ vhost_user_get_features(struct virtio_net **pdev,
+			struct vhu_msg_context *ctx __rte_unused,
+			int main_fd __rte_unused)
+{
+    VHOST_LOG_CONFIG(ERR, "vhost_user_set_config not supported");
```
in the commit msg you say, it is must, but in the actual code it does nothing.

How does it work?
When set_config() called?
Does the caller ignore the return status?

if so, can this routine be set to NULL?

What all config fields are setup today by QEMU on this area?

kailiangz1 replied:
set_config is read only for virtio-blk device, so, remove it.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -11,63 +11,19 @@
+#define VIRTIO_VDPA_INTR_RETRIES 256
+
+extern struct virtio_vdpa_device_callback virtio_vdpa_blk_callback;
+extern struct virtio_vdpa_device_callback virtio_vdpa_net_callback;
```
@kailiangz1 we have layering problem but can be fixed later.

paravmellanox commented:
Overall we need some improvements, but this is good enough start.
so lets revisit the module layering again in some time.
